src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1428 @@
+            private static BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern, TypeSymbol inputType)
+            {
+                if (pattern.InputType.Equals(inputType, TypeCompareKind.AllIgnoreOptions))
+                {
+                    return pattern;
+                }
+
+                if (pattern is BoundTypePattern typePattern1)
+                {
+                    return typePattern1.Update(typePattern1.DeclaredType, typePattern1.IsExplicitNotNullTest, inputType, typePattern1.NarrowedType);
+                }
+
+                if (pattern is BoundRecursivePattern recursivePattern)
+                {
+                    return recursivePattern.Update(
+                        recursivePattern.DeclaredType ??
+                            new BoundTypeExpression(recursivePattern.Syntax, aliasOpt: null, recursivePattern.InputType.StrippedType()),
+                        recursivePattern.DeconstructMethod, recursivePattern.Deconstruction,
+                        recursivePattern.Properties, recursivePattern.IsExplicitNotNullTest,
+                        recursivePattern.Variable, recursivePattern.VariableAccess,
+                        inputType, recursivePattern.NarrowedType);
+                }
+
+                if (pattern is BoundDiscardPattern discardPattern)
+                {
+                    return discardPattern.Update(inputType, inputType);
+                }
+
+                if (pattern is BoundNegatedPattern negatedPattern)
+                {
+                    return negatedPattern.Update(
+                        WithInputTypeCheckIfNeeded(negatedPattern.Negated, inputType),
+                        inputType, inputType);
+                }
+
+                Debug.Assert(pattern is BoundConstantPattern or BoundRelationalPattern or BoundITuplePattern or BoundListPattern);
+
+                // Produce `PatternInputType and pattern` given a new input type
+
+                BoundPattern typePattern = new BoundTypePattern(pattern.Syntax,
+                    new BoundTypeExpression(pattern.Syntax, aliasOpt: null, pattern.InputType),
+                    isExplicitNotNullTest: false, inputType, narrowedType: pattern.InputType).MakeCompilerGenerated();
+
+                var result = new BoundBinaryPattern(pattern.Syntax, disjunction: false, left: typePattern, right: pattern, inputType, pattern.NarrowedType);
+
+                if (pattern.WasCompilerGenerated)
+                {
+                    result = result.MakeCompilerGenerated();
+                }
+
+                return result;
+            }
+
+            // If the type of the declaration pattern implies a meaningful type test
+            // then we keep it.
+            // Otherwise, we push a discard.
+            public override BoundNode? VisitDeclarationPattern(BoundDeclarationPattern node)
+            {
+                TryPushOperand(NegateIfNeeded(MakeDiscardPattern(node).MakeCompilerGenerated()));
```

Comment:
>MakeDiscardPattern(node) [](http://example.com/codeflow?start=46&length=24)

Always creating a discard pattern feels suspicious. I assume that a declaration pattern might imply a meaningful type check. Using a discard instead is going to eliminate the type check and, therefore, have an undesired effect on the meaning of the resulting pattern.

Reply:
We might never want to report ```node``` as redundant, but there are other ways to achieve this without possibly loosing a meaningful type test and null test. For non-var, I am leaning towards  creating a compiler generated ``BoundTypePattern``` for declared type. We need to come up with a test scenario to back the behavior.